### PR TITLE
Np 46986 add import details

### PR DESCRIPTION
**Pull request by joachimjorgensen**
Add import details when creating a publication (resource) from Brage, Cristin or Scopus. 

Clarifications:
- ImportDetails for Scopus import is injected when the publication is created, not when raw data is imported.
- Same as above for Brage and Cristin, inject when creating the resource.

### CODE REVIEW

File: cristin-import/src/main/java/no/unit/nva/cristin/lambda/PublicationRepresentations.java [hunk 1]
```diff
@@ -94,11 +97,24 @@ public boolean cristinObjectIsPartOfAnotherPublication() {
     }
 
     public boolean updateHasEffectiveChanges() {
-        var existingPublication = getExistingPublication().copy().withModifiedDate(null).withIdentifier(null).build();
-        var incomingPublication = getIncomingPublication().copy().withModifiedDate(null).withIdentifier(null).build();
+        var existingPublication = getExistingPublication().copy()
+                                      .withModifiedDate(null)
+                                      .withIdentifier(null)
+                                      .withImportDetails(getConcatenatedImportDetailsFromExistingAndIncomingPublication())
```
Comment (joachimjorgensen) [inline cristin-import/src/main/java/no/unit/nva/cristin/lambda/PublicationRepresentations.java:103]: Made life difficult for myself by not allowing removal of import details

File: publication-commons/src/main/java/no/unit/nva/publication/service/impl/ResourceService.java [hunk 2]
```diff
@@ -163,13 +165,16 @@ public Publication createPublicationWithPredefinedCreationDate(Publication input
...
 
-    public Publication createPublicationFromImportedEntry(Publication inputData) {
+    public Publication createPublicationFromImportedEntry(Publication inputData, ImportSource importSource) {
         Resource newResource = Resource.fromPublication(inputData);
         newResource.setIdentifier(identifierSupplier.get());
         newResource.setPublishedDate(inputData.getPublishedDate());
         newResource.setCreatedDate(inputData.getCreatedDate());
         newResource.setModifiedDate(inputData.getModifiedDate());
         newResource.setStatus(PUBLISHED);
+        if (importSource != null) {
```
Comment (joachimjorgensen) [inline publication-commons/src/main/java/no/unit/nva/publication/service/impl/ResourceService.java:175]: Need this one until eventually CreatePublishedPublicationHandler is removed

File: publication-event-handlers/src/main/java/no/unit/nva/publication/events/handlers/create/CreatePublishedPublicationHandler.java [hunk 3]
```diff
@@ -68,7 +68,7 @@ private String randomUnitUser() {
     }
     
     private Publication storeAsPublishedPublication(Publication publication) {
-        return resourceService.createPublicationFromImportedEntry(publication);
+        return resourceService.createPublicationFromImportedEntry(publication, null);
```
Comment (joachimjorgensen) [inline publication-event-handlers/src/main/java/no/unit/nva/publication/events/handlers/create/CreatePublishedPublicationHandler.java:71]: Is this handler in use?

File: cristin-import/src/main/java/no/unit/nva/cristin/lambda/PublicationRepresentations.java (hunk 1 above)
Comment (truhacevkir) [reply cristin-import/src/main/java/no/unit/nva/cristin/lambda/PublicationRepresentations.java:103]: It seems like apache has a library for doing it, which is already in use in publication-api:

Try something like this:
`
!EqualsBuilder.reflectionEquals(existingPublication, incomingPublication, 
"modifiedDate", "identifier", "importDetails")
`

I think it is more elegant :)

**Review by truhacevkir — APPROVED** [review]
Looks good, single comment :)

File: publication-commons/src/main/java/no/unit/nva/publication/service/impl/ResourceService.java (hunk 2 above)
Comment (truhacevkir) [reply publication-commons/src/main/java/no/unit/nva/publication/service/impl/ResourceService.java:175]: can we use `nonNull()`

File: publication-event-handlers/src/main/java/no/unit/nva/publication/events/handlers/create/CreatePublishedPublicationHandler.java (hunk 3 above)
Comment (truhacevkir) [reply publication-event-handlers/src/main/java/no/unit/nva/publication/events/handlers/create/CreatePublishedPublicationHandler.java:71]: I don't think so. It seems like the handler, which managed the publication creation by scopus-import, which was located in the fetch-doi project when scopus-Import was using api calls to create the publications.

Can be removed.

File: cristin-import/src/main/java/no/unit/nva/cristin/lambda/PublicationRepresentations.java (hunk 1 above)
Comment (joachimjorgensen) [reply cristin-import/src/main/java/no/unit/nva/cristin/lambda/PublicationRepresentations.java:103]: Couldn't get EqualsBuilder to actually ignore specified fields... I'll let this one be unless someone has another smoother ways to deal with this.

File: publication-commons/src/main/java/no/unit/nva/publication/model/storage/ResourceDao.java [hunk 4]
```diff
@@ -153,6 +161,14 @@ public void setDoi(URI doi) {
         this.doi = doi;
     }
 
+    public List<ImportDetail> getImportDetails() {
+        return nonNull(importDetails) ? importDetails : Collections.emptyList();
+    }
+
+    public void setImportDetails(Collection<ImportDetail> importDetails) {
+        this.importDetails = new ArrayList<>(importDetails);
```
Comment (joachimjorgensen) [inline publication-commons/src/main/java/no/unit/nva/publication/model/storage/ResourceDao.java:169]: This one (and the other `setImportDetails`-setters in this repo) doesn't have the restriction that Publication has (from the datamodel). Here, one can remove existing `importDetails`. Not sure if this will be a problem.

File: brage-import/src/test/java/no/sikt/nva/brage/migration/lambda/BrageEntryEventConsumerTest.java [hunk 5]
```diff
@@ -1798,6 +1802,15 @@ void shouldUpdateExistingPublicationByFillingUpInstanceTypeEmptyValues(Publicati
                    doesNotHaveEmptyValuesIgnoringFields(Set.of(".pages.introduction", ".pages.illustrated")));
     }
 
+    @Test
+    void shouldAddImportDetailWhenImportingBrageRecord() throws IOException {
+        var brageGenerator = new NvaBrageMigrationDataGenerator.Builder().withType(TYPE_REPORT_WORKING_PAPER).build();
+        var s3Event = createNewBrageRecordEvent(brageGenerator.getBrageRecord());
+        var publication = handler.handleRequest(s3Event, CONTEXT).publication();
+        assertFalse(publication.getImportDetails().isEmpty());
```
Comment (brinxmat) [inline brage-import/src/test/java/no/sikt/nva/brage/migration/lambda/BrageEntryEventConsumerTest.java:1810]: This is necessarily false in the second assert

File: cristin-import/src/main/java/no/unit/nva/cristin/lambda/PublicationRepresentations.java (hunk 1 above)
Comment (brinxmat) [reply cristin-import/src/main/java/no/unit/nva/cristin/lambda/PublicationRepresentations.java:103]: Can ping me

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: brage-import/src/main/java/no/sikt/nva/brage/migration/lambda/BrageEntryEventConsumer.java [hunk 6]
```diff
@@ -334,7 +334,12 @@ private Try<PublicationRepresentation> tryPersistingInDatabase(
 
     private PublicationRepresentation createPublication(PublicationRepresentation publicationRepresentation) {
         var updatedPublication =
-            resourceService.createPublicationFromImportedEntry(publicationRepresentation.publication());
+            resourceService.createPublicationFromImportedEntry(publicationRepresentation.publication(),
+                                                               ImportSource.fromBrageArchive(
+                                                                   publicationRepresentation
+                                                                       .brageRecord()
+                                                                       .getCustomer()
+                                                                       .getName()));
```
Comment (joachimjorgensen) [inline brage-import/src/main/java/no/sikt/nva/brage/migration/lambda/BrageEntryEventConsumer.java:342]: The name of the archive of the Brage import

File: cristin-import/src/main/java/no/unit/nva/cristin/lambda/PublicationRepresentations.java (hunk 1 above)
Comment (joachimjorgensen) [reply cristin-import/src/main/java/no/unit/nva/cristin/lambda/PublicationRepresentations.java:103]: Conclusion: This method should be moved into the datamodel

File: cristin-import/src/test/java/no/unit/nva/cristin/lambda/CristinEntryEventConsumerTest.java [hunk 7]
```diff
@@ -1027,6 +1030,23 @@ void shouldUpdateNpiSubjectHeadingWhenMissing()
...
+        var publication = handler.handleRequest(sqsEvent, CONTEXT).getFirst();
+
+        var importDetail = publication.getImportDetails()
+                               .stream()
+                               .filter(f -> f.importSource().getSource().equals(Source.CRISTIN))
+                               .findFirst()
+                               .orElse(null);
+
+        assertNotNull(importDetail);
+        assertNull(importDetail.importSource().getArchive());
```
Comment (joachimjorgensen) [inline cristin-import/src/test/java/no/unit/nva/cristin/lambda/CristinEntryEventConsumerTest.java:1047]: Should add test to verify that `importDate` is not null?

File: publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/NonDegreePermissionStrategyTest.java [hunk 8]
```diff
@@ -165,7 +166,8 @@ void shouldAllowEmbargoThesisCuratorOnDegreeWithEmbargo()
     public static PublishedFile randomFileWithEmbargo() {
         return new PublishedFile(UUID.randomUUID(), RandomDataGenerator.randomString(),
                                  RandomDataGenerator.randomString(), RandomDataGenerator.randomInteger().longValue(),
-                                 RandomDataGenerator.randomUri(), false, true, Instant.now().plusSeconds(60 * 60 * 24),
+                                 RandomDataGenerator.randomUri(), false, PublisherVersion.PUBLISHED_VERSION,
```
Comment (joachimjorgensen) [inline publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/NonDegreePermissionStrategyTest.java:169]: Randomly had to update this one after bumping the datamodel 🤷

File: cristin-import/src/test/java/no/unit/nva/cristin/lambda/CristinEntryEventConsumerTest.java (hunk 7 above)
Comment (brinxmat) [reply cristin-import/src/test/java/no/unit/nva/cristin/lambda/CristinEntryEventConsumerTest.java:1047]: Probably?

**Review by brinxmat — APPROVED** [review]
